SG14/inplace_function.h

```diff
@@ -29,340 +29,308 @@
+  }
+
+  inplace_function& operator= (std::nullptr_t) noexcept
+  {
+    vtable_ptr_->destructor_ptr(std::addressof(storage_));
+    vtable_ptr_ = detail::empty_vtable_ptr<R, Args...>;
+    return *this;
+  }
+
+  inplace_function& operator= (const inplace_function& other)
+  {
+    if(this != std::addressof(other))
+    {
+      vtable_ptr_->destructor_ptr(std::addressof(storage_));
+
+      vtable_ptr_ = other.vtable_ptr_;
+      vtable_ptr_->copy_ptr(
+        std::addressof(storage_),
+        std::addressof(other.storage_)
+      );
+    }
+    return *this;
+  }
+
+  inplace_function& operator= (inplace_function&& other)
+  {
+    if(this != std::addressof(other))
+    {
+      vtable_ptr_->destructor_ptr(std::addressof(storage_));
+
+      vtable_ptr_ = other.vtable_ptr_;
+      vtable_ptr_->move_ptr(
+        std::addressof(storage_),
+        std::addressof(other.storage_)
+      );
+
+      other.vtable_ptr_ = detail::empty_vtable_ptr<R, Args...>;
+    }
+    return *this;
+  }
+
+  ~inplace_function()
+  {
+    vtable_ptr_->destructor_ptr(std::addressof(storage_));
+  }
+
+  R operator() (Args... args) const
+  {
+    return vtable_ptr_->invoke_ptr(
+      std::addressof(storage_),
+      std::forward<Args>(args)...
+    );
+  }
+
+  constexpr bool operator== (std::nullptr_t) const noexcept
+  {
+    return !operator bool();
+  }
+
+  constexpr bool operator!= (std::nullptr_t) const noexcept
 	{
 		return operator bool();
 	}
 
-	// Converts to 'true' if assigned
-	explicit constexpr operator bool() const noexcept
+  explicit constexpr operator bool() const noexcept
 	{
-		return m_InvokeFctPtr != &DefaultFunction;
+	  return vtable_ptr_ != detail::empty_vtable_ptr<R, Args...>;
 	}
 
-	// Invokes the target
-	// Throws std::bad_function_call if not assigned
-	RetT operator () (ArgsT... args) const
-	{
-		return m_InvokeFctPtr(std::forward<ArgsT>(args)..., data());
-	}
-
-	// Swaps two targets
-	void swap(inplace_function& other)
-	{
-		BufferType tempData;
-		this->move(m_Data, tempData);
-		other.move(other.m_Data, m_Data);
-		this->move(tempData, other.m_Data);
-		std::swap(m_InvokeFctPtr, other.m_InvokeFctPtr);
-		std::swap(m_ManagerFctPtr, other.m_ManagerFctPtr);
-	}
+  template<size_t Cap, size_t Align>
+  operator inplace_function<R(Args...), Cap, Align>() const
+  {
+    static_assert(detail::is_valid_inplace_dst<
+      Cap, Align, Capacity, Alignment
+    >::value);
+
+    std::aligned_storage_t<Cap, Align> conv_storage;
+    vtable_ptr_->copy_ptr(
+      std::addressof(conv_storage),
+      std::addressof(storage_)
+    );
+    return {vtable_ptr_, std::move(conv_storage)};
+  }
+
+  // not sure a portable memcopy && conversion solution exists
+
+  void swap(inplace_function& other)
+  {
+    if (this == std::addressof(other)) return;
+
+    storage_t tmp;
+    vtable_ptr_->move_ptr(
+      std::addressof(tmp),
+      std::addressof(storage_)
+    );
+    vtable_ptr_->move_ptr(
```

Comment:
I am fairly confident that at least one of these `vtable_ptr_`s should be an `other.vtable_ptr_`.

Reply:
Test case filed as #117. It turns out that master is also broken here, but in a different (and possibly worse) way; #117 contains a fix for master as well as the new test case.